### ISSUE TITLE
how to properly use .build() method

### ISSUE THREAD

**Issue by kirk86** [fchollet/keras#5342]
Hi guys, 
I was wondering what's the proper usage of `.build()` method. For instance I wanted to accomplish something like the following:

```
from keras.applications import VGG16
from keras.layers.core import Dense

model = VGG16(include_top=True, weights=None)
model.layers.pop()
model.layers.append(Dense(10, activation='softmax', name='predictions'))
model.summary()
```

but `model.summary()` throws error about layer not being built, so I tried the following

```
model.build(model.get_layer(name='fc2').output_shape)
model.summary()
```
still throws error about layer not being built. I'm lost on how to use `.build()` to build that particular layer.

The above way of doing things feel more natural and I think the `.build()` should be called automatically when the user uses `model.layers.append()` instead of having every time to call it explicitly. 

Furthermore, the above code seems sorter and more readable than using the following functional api approach

```
rom keras.applications import VGG16
from keras.layers.core import Dense
from keras.layers import Model

model = VGG16(include_top=True, weights=None)
model.layers.pop()
x = model.output
x = Dense(10, activation='softmax', name='predictions')(x)
model = Model(model.input , x)
model.compile(...)
model.summary()
```

**Comment by unrealwill** [comment]
I'm not familiar at all with using build() to construct the network. Though I remember doing something in this style with the Lasagne library (which you may have heard of). I remember it because I got burned by it, and moved to Keras. 

It seems build is designed to be idempotent (You can call it as many time as you want and it won't change the result).

The functional API is clean, I recommend using it, I guess it's what the rest is build upon.
I guess to call build() properly you would have to do what is done in 
Layer "__call__"
line 507: https://github.com/fchollet/keras/blob/master/keras/engine/topology.py 

Using the functional API will also allow to reuse weights.

**Comment by kirk86** [comment]
@unrealwill thanks for the answer. What I meant is that it seems more natural when you start with predefined networks to pop as many layers as you'd like and append as many as you want without having errors when you print the summary. The other way around it is to use the functional api but it would be sorted and cleaner to simply say pop `x`  layers append `y` layers and be done with it. No need for external call to `.build()` or anything like that. That would be taken care of by `.compile()` for instance.

**Comment by unrealwill** [comment]
If you really like your style more, you can write custom function for the operation you usually do on Models
Like
def PopLastLayerFromModel( model ):
    use functionnal API here
    return newModel

def AppendCusomLayerToModel( model, layer ):
   use functionnal API here
   return newModel
   
I mean build() is not part of the functionnal API:
https://keras.io/models/model/

So when you are modifying layers, you are modifying internals, which means it's not guaranteed to be preserved in the furure in the way you use it, it may be broken by anyone.

Because Keras is in python it relies on the programmer responsibility to respect public interfaces. In other languages, we would have some private or protected attributes.

**Comment by kirk86** [comment]
@unrealwill Yup, I got your point  but my initial question still remains even by looking at build used int the `topology.py` file. For instance I would like to know what I've done wrong in the example in my first question where I provide a case study of using `build()`. That example throws an error. What's the proper way to use it in the frame of that minimal concept example I've provided.

**Comment by piyush05124** [comment]
is there any perfect solution to this issue
